sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -53,7 +53,15 @@ import org.apache.spark.util.Utils
 
 private[sql] object Dataset {
   def apply[T: Encoder](sparkSession: SparkSession, logicalPlan: LogicalPlan): Dataset[T] = {
-    new Dataset(sparkSession, logicalPlan, implicitly[Encoder[T]])
+    val encoder = implicitly[Encoder[T]]
+    if (encoder.clsTag.runtimeClass == classOf[Row]) {
+      // We should use the encoder generated from the executed plan rather than the existing
+      // encoder for DataFrame because the types of columns can be varied due to widening types.
+      // See SPARK-17123. This is a bit hacky. Maybe we should find a better way to do this.
+      ofRows(sparkSession, logicalPlan).asInstanceOf[Dataset[T]]
+    } else {
+      new Dataset(sparkSession, logicalPlan, encoder)
+    }
```

Comment:
Ah, the reason why `intersect` was fine seems because it does contains any data. `intersect` seems also failed:

```
val dates = Seq(Tuple1(BigDecimal.valueOf(10.5))).toDF("decimal")
val widenTypedRows =  Seq(Tuple1(10.5D)).toDF("decimal")
dates.intersect(widenTypedRows).collect()
```